### ISSUE TITLE
Dot in li element is not in the middle

### ISSUE THREAD

**Issue by mndzielski** [danfickle/openhtmltopdf#478]
Hi,

I'd like to report that dot in li element is align to top. In browser is align to middle. It's quite annoying.

Another issue is fact that dot is not fully rendering when font-size is large.

HTML example:

```<html>
<head>
<style>
#e1 {
 font-size: 40px;
}

#e2 {
 font-size: 55px;
}
</style>
</head>
<body>

<ul>
<li id="e1">Hello world</li>
<li id="e2">Hello world</li>
</ul>

</body>
</html>
```

And result:

![image](https://user-images.githubusercontent.com/38613340/81703840-62e8ee00-946d-11ea-888d-34e88679f0de.png)

**Comment by syjer** [comment]
hi @mndzielski , I'll have a look

**Comment by mndzielski** [comment]
@danfickle @syjer 
Is there any update on this issue?
![image](https://user-images.githubusercontent.com/38613340/91462740-d2eb7a80-e88a-11ea-82ee-d1e665bf4edc.png)

**Comment by syjer** [comment]
hi @mndzielski currently not, I was side tracked on other stuff, but I'll have look in this period.

**Comment by mndzielski** [comment]
Ok,
so my temporary solution for that issue is:

`
ul {
            list-style: none;
        }
`

` 
        ul li::before {
            content: "\2022";
            font-weight: bold;
            display: inline-block;
            width: 1em;
            margin-left: -1em;
        }`

Hope this may help others

![image](https://user-images.githubusercontent.com/38613340/91463794-198da480-e88c-11ea-8741-25daebfe7e38.png)

**Comment by syjer** [comment]
Currently trying to understand the issue of vertical positioning:

looking at the code in `ListItemPainter.java`:

```java
int y = getReferenceBaseline(c, box) - (int)strutMetrics.getAscent() / 2  - marker.getDiameter() / 2;
```

seems to skew a little bit too much on the top.

Adding some code to visualize it:

```java
c.getOutputDevice().drawDebugOutline(c, box, FSRGBColor.GREEN);
c.getOutputDevice().setColor(FSRGBColor.RED);
c.getOutputDevice().drawRect(x, getReferenceBaseline(c, box), 1000, 1);
c.getOutputDevice().setColor(FSRGBColor.BLUE);
c.getOutputDevice().drawRect(x, getReferenceBaseline(c, box) - (int)strutMetrics.getAscent(), 1000, 1);
```

We can see the issue of the vertical positioning:

![Screenshot from 2020-08-27 18-45-13](https://user-images.githubusercontent.com/498146/91470930-a1c47780-e895-11ea-9640-a3ab698af7b8.png)

Basically, we center exactly with the full line height, which is visually not exactly correct.

**Comment by syjer** [comment]
For the horizontal alignement:

```java
if (style.getDirection() == IdentValue.LTR) {
            x += -marker.getLayoutWidth() + marker.getDiameter();
        }
```

Adding `+ marker.getDiameter();` seems to do the trick

**Comment by syjer** [comment]
doing this in `ListItemPainter.drawGlyph`:

```java
        if (style.getDirection() == IdentValue.LTR) {
            x += -marker.getLayoutWidth() + marker.getDiameter();
        }
        if (style.getDirection() == IdentValue.RTL){
            x += box.getMarkerData().getReferenceLine().getWidth() + marker.getLayoutWidth();
        }

        //

        c.getOutputDevice().drawDebugOutline(c, box, FSRGBColor.GREEN);

        c.getOutputDevice().setColor(FSRGBColor.RED);
        c.getOutputDevice().drawRect(x, getReferenceBaseline(c, box), 1000, 1);


        c.getOutputDevice().setColor(FSRGBColor.BLUE);


        int bottomLine = getReferenceBaseline(c, box);
        int top = bottomLine - (int) (strutMetrics.getAscent() / 1.5);
        c.getOutputDevice().drawRect(x, top, 1000, 1);
        //

        c.getOutputDevice().setColor(new FSRGBColor(255, 255, 0));
        c.getOutputDevice().drawRect(x, bottomLine - (bottomLine-top)/2, 1000, 1);
        //


        c.getOutputDevice().setColor(FSRGBColor.BLACK);
        int y = bottomLine - (bottomLine-top) /2 - marker.getDiameter() / 2 ;
```

Using 1.5 as a criteria for finding the "top" of the line does not seems to bad as a heuristic.

we get 
![Screenshot from 2020-08-27 19-11-22](https://user-images.githubusercontent.com/498146/91473548-514f1900-e899-11ea-9732-819f42f2c3d2.png)

which is quite decent.

l'll need to test with the RTL and look if there are any corner case.

WDYT @danfickle ?

(looking at the failing  tests, the placement of the disc seems to be a little better now:

before:
![columns-nested-unbalanced---1---expected](https://user-images.githubusercontent.com/498146/91475148-76448b80-e89b-11ea-9f77-9b4d08519bf3.png)

after:

![columns-nested-unbalanced---1---actual](https://user-images.githubusercontent.com/498146/91475162-7b093f80-e89b-11ea-82ba-a45f91fa8775.png)

)

**Comment by danfickle** [comment]
@syjer, I'm happy to go with your judgement on this one. It does look better with your heuristic. I'm sure you're well ahead of me, but remember to test(at least informally) with at least a couple of fonts and font sizes.

Thanks again.

**Comment by syjer** [comment]
@danfickle yes, I'm currently looking at the RTL case (and adding some fine tuning for the x positioning for the LTR one)

I've noticed that the list (ul, ol) in chrome&co have a css property "padding-inline-start: 40px" which, in function if it's a LTR or RTL case, will add the necessary padding on the right or left.

As far as I've noticed, we don't really support this property, as we define in our default css:

```
ol, ul, dir,
menu            { padding-left: 40px }
```

Which work really well for LTR, but clearly not for RTL.

So I'm wondering if I should fix first the LTR case and submit the PR, and then try to tackle the RTL one which seems more complex, as we need to support a new css property.
